### PR TITLE
8253524: C2: Refactor code that clones predicates during loop unswitching

### PR DESCRIPTION
**Pull request by rwestrel**
During unswitching, PhaseIdealLoop::create_slow_version_of_loop()
calls PhaseIdealLoop::clone_predicates_to_unswitched_loop() twice, one
for each loops, to clone some predicates above each loop. That code is
fragile as it (implicitly) requires the fast loop to be processed
first. I propose calling
PhaseIdealLoop::clone_predicates_to_unswitched_loop() a single time
and have it handle both loops in a single pass.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253524](https://bugs.openjdk.java.net/browse/JDK-8253524): C2: Refactor code that clones predicates during loop unswitching

### Reviewers
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 5ab947eba0b0a1f87cf3bfa232915a80722c7a35
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to 5ab947eba0b0a1f87cf3bfa232915a80722c7a35

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/317/head:pull/317`
`$ git checkout pull/317`

### CODE REVIEW

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 1]
```diff
@@ -225,17 +224,14 @@ ProjNode* PhaseIdealLoop::clone_predicate_to_unswitched_loop(ProjNode* predicate
...
 }
 
 // Clones skeleton predicates starting at 'old_predicate_proj' to
 // 'new_predicate_proj' and rewires the control edges of data nodes in
 // the loop from the old predicates to the new cloned predicates.
-void PhaseIdealLoop::clone_skeleton_predicates_to_unswitched_loop(Deoptimization::DeoptReason reason, ProjNode* old_predicate_proj,
-                                                    ProjNode* new_predicate_proj, bool is_slow_loop, uint idx_before_clone,
-                                                    Node_List &old_new) {
-  assert(old_predicate_proj->is_Proj(), "must be projection");
+void PhaseIdealLoop::clone_skeleton_predicates_to_unswitched_loop(IdealLoopTree *loop, const Node_List &old_new, Deoptimization::DeoptReason reason,
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopPredicate.cpp:233]: Asterisk: `IdealLoopTree*`

File: src/hotspot/share/opto/loopnode.hpp [hunk 2]
```diff
@@ -1431,13 +1431,11 @@ class PhaseIdealLoop : public PhaseTransform {
...
-  ProjNode* clone_predicate_to_unswitched_loop(ProjNode* predicate_proj, Node* new_entry, Deoptimization::DeoptReason reason,
-                                 bool is_slow_loop, uint idx_before_clone, Node_List &old_new);
-  void clone_skeleton_predicates_to_unswitched_loop(Deoptimization::DeoptReason reason, ProjNode* old_predicate_proj,
-                                      ProjNode* new_predicate_proj, bool is_slow_loop,
-                                      uint idx_before_clone, Node_List &old_new);
+  void clone_predicates_to_unswitched_loop(IdealLoopTree *loop, const Node_List &old_new, ProjNode*& iffast, ProjNode*& ifslow);
+  ProjNode* clone_predicate_to_unswitched_loop(ProjNode* predicate_proj, Node* new_entry, Deoptimization::DeoptReason reason);
+  void clone_skeleton_predicates_to_unswitched_loop(IdealLoopTree *loop, const Node_List &old_new, Deoptimization::DeoptReason reason,
+                                      ProjNode* old_predicate_proj, ProjNode* iffast, ProjNode* ifslow);
+  void check_created_predicate_for_unswitching(const Node *new_entry) const PRODUCT_RETURN;
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.hpp:1438]: Move the asterisks to the types.

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 3]
```diff
@@ -349,39 +318,45 @@ Node* PhaseIdealLoop::clone_predicates_to_unswitched_loop(Node* old_entry, Node*
...
+    ifslow = clone_predicate_to_unswitched_loop(limit_check_proj, ifslow, Deoptimization::Reason_loop_limit_check);
+
+    check_created_predicate_for_unswitching(iffast);
+    check_created_predicate_for_unswitching(ifslow);
   }
-  return new_entry;
 }
 
+#ifndef PRODUCT
+void PhaseIdealLoop::check_created_predicate_for_unswitching(const Node *new_entry) const {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopPredicate.cpp:350]: Asterisk: `Node*`

**Review by chhagedorn — CHANGES_REQUESTED** [review]
Very nice clean-up! This makes it much easier. Apart from some minor code style comments, it looks good to me!

File: src/hotspot/share/opto/loopPredicate.cpp (hunk 3 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopPredicate.cpp]: @chhagedorn thanks for the review (and offline comments). Should be fixed now.

**Review by chhagedorn — APPROVED** [review]
Looks good to me!

**Review by vnkozlov — APPROVED** [review]

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 4]
```diff
@@ -279,58 +265,41 @@ void PhaseIdealLoop::clone_skeleton_predicates_to_unswitched_loop(Deoptimization
...
-#ifdef ASSERT
-  assert(LoopUnswitching, "sanity - only called when unswitching a loop");
-  if (new_entry == NULL || !(new_entry->is_Proj() || new_entry->is_Region() || new_entry->is_SafePoint())) {
-    if (new_entry != NULL)
-      new_entry->dump();
-    assert(false, "not IfTrue, IfFalse, Region or SafePoint");
-  }
-#endif
+void PhaseIdealLoop::clone_predicates_to_unswitched_loop(IdealLoopTree* loop, const Node_List& old_new, ProjNode*& iffast, ProjNode*& ifslow) {
+  LoopNode* head  = loop->_head->as_Loop();
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopPredicate.cpp:298]: Extra whitespace after `head`

**Review by TobiHartmann — APPROVED** [review]
Looks good.

File: src/hotspot/share/opto/loopPredicate.cpp (hunk 4 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopPredicate.cpp:298]: I'll remove the white space before I integrate the change.

**Comment by rwestrel** [comment]
@chhagedorn @vnkozlov @TobiHartmann thanks for the review

**Review by chhagedorn — APPROVED** [review]

**Pull request closed without merging**